### PR TITLE
Partial fix of Issue 8384 - std.conv.to should allow conversion betwe…

### PR DESCRIPTION
**Pull request by Biotronic**
…en any pair of string/wstring/dstring/char*/wchar*/dchar*

std.conv.to currently fails uglily with 'object.Error@(0): array cast misalignment.' when casting from a char* to any wider string type, and does not support any kind of conversion from e.g. a wchar* to wchar[]. This should fix both those issues.

Note that Issue 8384 also mentions conversions from {X}char[] to {Y}char*. I have not tackled that problem, as std.conv.to does not currently support any kind of conversion in that direction.

### CODE REVIEW

File: std/conv.d [hunk 1]
```diff
@@ -875,11 +875,19 @@ T toImpl(T, S)(S value)
         ()@trusted{ memcpy(result.ptr, value.ptr, value.length); }();
         return cast(T) result;
     }
-    else static if (isPointer!S && is(S : const(char)*))
+    else static if (isPointer!S && (is(S : const(char)*) || is(S : const(wchar)*) || is(S : const(dchar)*)))
```
Comment (burner) [inline std/conv.d]: the rest of the constraint should use std.traits.isSomeChar

File: std/conv.d [hunk 2]
```diff
@@ -875,11 +875,19 @@ T toImpl(T, S)(S value)
...
     {
-        import core.stdc.string : strlen;
-        // It is unsafe because we cannot guarantee that the pointer is null terminated.
-        return value ? cast(T) value[0 .. strlen(value)].dup : null;
+        // This is unsafe because we cannot guarantee that the pointer is null terminated.
```
Comment (burner) [inline std/conv.d]: does that code compile in a @safe unittest? if so. I would suggest making the block @system to hint at the danger

File: std/conv.d [hunk 3]
```diff
@@ -930,6 +938,30 @@ unittest
...
+            test2!(immutable(e)[])(lp);
+        }
+    }
+
+    void test2(D, S)(S lp)
+    {
+        assert(to!string(to!D(lp)) == "Hello, world!");
+    }
+
+    foreach (e; AliasSeq!("Hello, world!", "Hello, world!"w, "Hello, world!"d))
```
Comment (burner) [inline std/conv.d]: please also test an empty string

File: std/conv.d (hunk 2 above)
Comment (Biotronic) [inline std/conv.d]: No it doesn't. That's actually what's causing the auto-tester to fail win32 - std.datetime does std.conv.to!string(wchar*). It's obvious the code in std.datetime has not been tested, since to!string(wchar*) did *the wrong thing* until this PR.

That said, it's not actually less safe than what was done before.

Comment (burner) [inline std/conv.d]: I think I was not clear. This code should not be safe it should be marked system. So people using it must make a conscious to use this code.

Comment (Biotronic) [inline std/conv.d]: You were perfectly clear. Attribute inference marks it @system, and attempting to use it from a safe context will fail. std.datetime is attempting to use it from a safe context, and fails.

Comment (burner) [inline std/conv.d]: then I was not clear @system should not be inferred it should be explicit.

Comment (Biotronic) [inline std/conv.d]: That would require factoring out that part into a separate function. The other parts of the function are not @system, and due to templates they will not be marked @system if the instantiated part doesn't do anything shady. This is exactly what attribute inference is for.

File: std/conv.d (hunk 1 above)
Comment (Biotronic) [inline std/conv.d]: isSomeChar checks if it's a char of some sort, not a char* of some sort.

File: std/conv.d (hunk 2 above)
Comment (Biotronic) [inline std/conv.d]: Also, marking it @system will not force people to make a conscious choice. The choice they make is they look at the documention of std.conv, see it has a function called 'to', and then... No. That's it. They don't look at the source code to see what attributes are spattered across its internal workings.

Comment (burner) [inline std/conv.d]: in phobos there is an idioms of directly call @trusted delegates. you could make a @system delegate in place.
if you look at the error message currently generated you still have to look into "to" to find out what is not safe or trusted. And for that you have to understand the code. Which means work. But if you see "@system" you don't have to understand the code anymore

File: std/conv.d (hunk 1 above)
Comment (burner) [inline std/conv.d]: isSomeChar!(ElementEncodingType!(S))

File: std/conv.d (hunk 2 above)
Comment (Biotronic) [inline std/conv.d]: That's acceptable. Thanks.

Comment (burner) [inline std/conv.d]: sure, I would say that is code reviews are for

File: std/conv.d (hunk 1 above)
Comment (Biotronic) [inline std/conv.d]: Thankies.

File: std/datetime.d [hunk 4]
```diff
@@ -29929,7 +29929,7 @@ else version(Windows)
         }
 
 
-        this(string name, TIME_ZONE_INFORMATION tzInfo) @safe immutable pure
+        this(string name, TIME_ZONE_INFORMATION tzInfo) @trusted immutable pure
```
Comment (Biotronic) [inline std/datetime.d]: I'm more than a little unsure about this. The conversion on line 29934 is definitely wrong as is (conversion from wchar* to string does not work, and fails in ugly ways at runtime). The question is, should this function be @trusted? @system?
Should it even be using this conversion? Am I right to address the issue here or should that be a separate pull?

Regarding to the safety of using to!string(wchar*), MSDN claims in one place that DaylightName will always be zero-terminated. If this is true, this should be safe. However, there are many versions of the info page on TIME_ZONE_INFORMATION, and not all of them include this information. The constructor is private though, and I haven't found a case where it could be called with invalid data.

Comment (jmdavis) [inline std/datetime.d]: I believe that the consensus at this point is that if the @safety depends on the argument being correct, then it should be marked as `@trusted`. The reality of the matter in this case is that either will work. The constructor is private, and the function calling it is already `@trusted`. So, just mark it as `@trusted`.

And if the Windows API gives a `wchar*` that _isn't_ zero-terminated, then it's just plain broken, since it's a C API. But actually, in this case, it doesn't even matter, because the `TIME_ZONE_INFORMATION` is constructed by std.datetime in the caller, and it's explicitly zero-terminated.

Comment (Biotronic) [inline std/datetime.d]: Great, thanks.

**Comment by DmitryOlshansky** [comment]
LGTM

**Comment by DmitryOlshansky** [comment]
Auto-merge toggled on
